File: downloads/code/LeetCode/Java/Different-Ways-to-Add-Parentheses.java
public class Solution {
    public List<Integer> diffWaysToCompute(String input) {
        List<Integer> ret = new LinkedList<Integer>();
        for(int i=0;i<input.length();i++){
            if(input.charAt(i)=='+' || input.charAt(i)=='*' ||
               input.charAt(i)=='-'){
               
               List<Integer> left = diffWaysToCompute(input.substring(0,i));
               List<Integer> right = diffWaysToCompute(input.substring(i+1));

               for(int p1 : left){
                   for(int p2 : right){
                        int c = 0;
                        if(input.charAt(i) == '+')
                            c = p1+p2;
                        if(input.charAt(i) == '*')
                            c = p1*p2;
                        if(input.charAt(i) == '-')
                            c = p1-p2;
                   
                        ret.add(c);
                   }
               }
            }

        }       
        if(ret.size()==0)
            ret.add(Integer.valueOf(input));
        return ret;
    }
}

File: downloads/code/LeetCode/Python/Different-Ways-to-Add-Parentheses.py
class Solution(object):
    def cal(self, a,b,op):
        if op == "+":
            return a+b
        elif op == "-":
            return a-b
        else:
            return a*b

    def dfs(self, nums, ops):
        if not ops:
            return [nums[0]]
        
        # find split the nums and merge
        ans = []
        for op in range(len(ops)):
            left = self.dfs(nums[:op+1],ops[:op])
            right = self.dfs(nums[op+1:],ops[op+1:])
            for a in left:
                for b in right:
                    ans.append(self.cal(a,b,ops[op]))

        return ans

    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        operator = "+-*"
        nums = []
        ops = []

        # split input
        s = re.split('(\D)', input)
        for i in s:
            if i in operator:
                ops.append(i)
            else:
                nums.append(int(i))

        return self.dfs(nums,ops)
